qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py

```diff
@@ -0,0 +1,210 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2022 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+#
+from qtpy.QtWidgets import QVBoxLayout, QWidget, QTableWidget, QHeaderView, QTableWidgetItem
+from qtpy.QtCore import Qt
+from qtpy.QtGui import QColor
+from mantid.plots.plotfunctions import create_subplots
+from mantidqt.MPLwidgets import FigureCanvas
+from mantidqt.plotting.mantid_navigation_toolbar import MantidNavigationToolbar
+import matplotlib.text as text
+from mantid.simpleapi import AnalysisDataService as ADS
+from mantid.kernel import SpecialCoordinateSystem
+from numpy import zeros
+
+# local imports
+from .representation.cut_representation import CutRepresentation
+
+
+class CutViewerView(QWidget):
+    """Displays a table view of the PeaksWorkspace along with controls
+    to interact with the peaks.
+    """
+    def __init__(self, presenter, canvas, frame, parent=None):
+        """
+        :param painter: An object responsible for drawing the representation of the cut
+        :param sliceinfo_provider: An object responsible for providing access to current slice information
+        :param parent: An optional parent widget
+        """
+        super().__init__(parent)
+        self.presenter = presenter
+        self.layout = None
+        self.figure_layout = None
+        self.table = None
+        self.figure = None
+        self.cut_rep = None
+        self.canvas = canvas
+        self.frame = frame
+        self._setup_ui()
+        self._init_slice_table()
+        self.table.cellChanged.connect(self.on_cell_changed)
+
+    def hide(self):
+        super().hide()
+        if self.cut_rep is not None:
+            self.cut_rep.remove()
+            self.cut_rep = None
+
+    # signals
+
+    def on_cell_changed(self, irow, icol):
+        self.set_bin_params(*self.presenter.validate_bin_params(irow, icol))
+        self.presenter.update_cut()
+
+    def resizeEvent(self, event):
+        super().resizeEvent(event)
+        self.figure.tight_layout()
+
+    # getters
+
+    def get_step(self, irow):
+        return float(self.table.item(irow, 6).text())
+
+    def get_bin_params(self):
+        vectors = zeros((3, 3), dtype=float)
+        extents = zeros((2, 3), dtype=float)
+        nbins = zeros(3, dtype=int)
+        for ivec in range(vectors.shape[0]):
+            for icol in range(vectors.shape[0]):
```

Comment:
I'm confused why this is iterating through `shape[0]` twice - should the second one be `shape[1]`?

Reply:
I could have used `shape[1]`, but in this case it doesn't matter as `vectors` is a 3x3 matrix and always will be even for 2D and 4D workspaces - we will always have two vectors in the plane of the slice and one out of the plane of the slice.